content/collective-communication-pt1.rst

```diff
@@ -3,55 +3,126 @@ Collective communication 1
+any of them proceeds past it.
 
-A code block with preceeding paragraph::
 
-  import multiprocessing
+.. figure:: img/MPI_Barrier.svg
+   :align: center
 
-* A bullet list
+   All ranks in the communicator reach the barrier before any continue past it
 
-* Bullet list
+``MPI_Barrier`` is `blocking` (ie. does not return until the operation
+is complete) and introduces `collective synchronization` into the
+program. This can be useful to allow rank to wait for an external
+event (e.g. files being written by another process) before entering
+the barrier, rather than have all ranks checking.
 
-  * Sub-list::
+When debugging problems in other MPI communication, adding calls to
+``MPI_Barrier`` can be useful. However, if a barrier is necessary for
+your program to function correctly, that may suggest your
+program has a bug!
 
-      code block (note indention)
+Call signature
+~~~~~~~~~~~~~~
 
-  .. note::
+Call signature::
 
-     directive within a list (note indention)
+  int MPI_Barrier(MPI_Comm comm)
 
+Link to `MPI_Barrier man page <https://www.open-mpi.org/doc/v4.0/man3/MPI_Barrier.3.php>`_
 
-Exercise: the general topic
----------------------------
+Link to `Specification of MPI_Barrier <https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node100.htm#Node100>`_
 
-Exercises get their own section, so that they can be linked and found
-in the table of contents.
+.. note::
 
-.. challenge:: 1.1 Exercise title
+   It takes one argument, the communicator over which the barrier
+   operates.  All ranks within that communicator must call it or the
+   program will hang waiting for them to do so.
+
+
+Broadcast
+---------
+
+An ``MPI_Bcast`` call sends data from one rank to all other ranks.
+
+
+.. figure:: img/MPI_Bcast.svg
+   :align: center
+
+   After the call, all ranks in the communicator agree on the two values
+   sent.
+
+``MPI_Bcast`` is `blocking` and introduces `collective
+synchronization` into the program.
+
+This can be useful to allow one rank to share values to all
+other ranks in the communicator. For example, one rank might read
+a file, and then broadcast the content to all other ranks. This is
+usually more efficient than having each rank read the same file.
+
+Call signature::
+
+  int MPI_Bcast(void *buffer, int count, MPI_Datatype datatype,
+                int root, MPI_Comm comm)
+
+Link to `MPI_Bcast man page <https://www.open-mpi.org/doc/v4.0/man3/MPI_Bcast.3.php>`_
+
+Link to `Specification of MPI_Bcast <https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node101.htm#Node101>`_
+
+.. note::
+
+   All ranks must supply the same value for ``root``, which specifies
+   the rank of that communicator that provides the values that are
+   broadcast to all other ranks.
+
+   ``buffer``, ``count``, ``datatype``, and ``comm`` are similar to those
+   used for point-to-point communication; all ranks in the communicator
+   must participate with valid buffers and consistent counts and types.
```

Comment:
It's a good idea to separate the explanation of the arguments into a note, so they get highlighted. I will do the same in my episodes.